### ISSUE TITLE
Use explicit canonicalized JSON 

### ISSUE THREAD

**Issue by joepio** [joepio/atomic#115]
Although the tests are currently passing, and most commits seem to work just fine, I'm pretty sure that there will be situations where the deterministic json-ad serialization will be different from the Typescript version.

To prevent this, I should switch to [`serde_jcs`](https://github.com/l1h3r/serde_jcs). Also, double check in atomic-data-browser that this spec is used as well.

**Comment by joepio** [comment]
Also, don't use base64 in URLs (because of `=`, `/` characters), maybe consider https://tools.ietf.org/html/rfc4648

**Comment by joepio** [comment]
I used `serde_canonical_json`. Not sure it is actually JCS, though: https://github.com/gelvinp/rs-serde_canonical_json/issues/1

**Comment by joepio** [comment]
The JS one isn't officially JCS, but it _might_ be... 

https://github.com/epoberezkin/fast-json-stable-stringify/issues/79

If it turns out not to be this, we should switch to some other NPM JCS serialization tool, like:

- [canonicalize](https://github.com/erdtman/canonicalize), the oldest, 8yrs. Most downloads.
- [json-canonicalize](https://github.com/snowyu/json-canonicalize.ts), 7yrs old. Has some typescript. But[ I don't think its JCS](https://github.com/snowyu/json-canonicalize.ts/issues/23).
- [canonical-json](https://github.com/mirkokiefer/canonical-json), the youngest one, 2 months. But [probably not JCS](https://github.com/mirkokiefer/canonical-json/issues/14).

I [asked the authors for differences](https://github.com/erdtman/canonicalize/issues/15).

**Comment by jonassmedegaard** [comment]
Seems [serde_json_canonicalizer](https://lib.rs/crates/serde_json_canonicalizer) is a perfect fit.

**Comment by joepio** [comment]
@jonassmedegaard yeah that seems good for rust, but  I think our serde_jcs is fine. However, the issue now is with the node implementation - it's unclear whether that one is truly JCS. So far it seems like it though.

**Comment by midiakiasat** [comment]
One missing guardrail to fully close this:

You need **cross-language conformance tests**, not just library selection. JCS compliance must be asserted by shared, versioned test vectors (JSON input → canonical bytes → hash/signature) executed in **both Rust and Node**. If the Node implementation is “probably JCS” but unverified, determinism is still undefined at the system boundary.

Recommendation: publish a small JCS test corpus in-repo and gate CI on byte-for-byte equality across implementations. Without this, signatures may validate intra-language but diverge cross-stack, which defeats the purpose of canonicalization.

**Comment by mirkokiefer** [comment]
As of v0.3.0 (just released), canonical-json is RFC 8785 (JCS) compliant - https://github.com/mirkokiefer/canonical-json/commit/9e4a9c9821c175efc52c16343a4bf05a4985aac6

  - Object keys sorted by UTF-16 code unit order
  - Numbers serialized per ES6 Number.toString() rules (-0 → 0)
  - Only mandatory characters escaped (U+0000–U+001F, ", \)

We also ship a walk() function for streaming canonical chunks to a hasher without allocating the full string, and a convenience hash() via canonical-json/hash.

  Re @midiakiasat's point about cross-language conformance — happy to add the RFC 8785 test vectors to our CI if anyone has a shared corpus. That's a great idea.

**Comment by joepio** [comment]
Thank you @mirkokiefer :)
